### PR TITLE
Refactor crypto

### PR DESCRIPTION
**Pull request by roman-khimov**
This moves some functionality into micro-packages, improves testing, unexports some code and fixes bugs along the way.

### CODE REVIEW

File: pkg/encoding/base58/base58.go [hunk 1]
```diff
@@ -16,13 +16,6 @@ func CheckDecode(s string) (b []byte, err error) {
 		return nil, err
 	}
 
-	for i := 0; i < len(s); i++ {
```
Comment (fyrchik) [inline pkg/encoding/base58/base58.go]: This is needed when base58 is used to encode a number where `001 == 1`. `1` are used to encode leading `0`.
`mr-tron`'s implementation does this by default, so you have actually fixed a bug.

This test fails on master, can you add it?
```golang
func TestBase58LeadingZeroes(t *testing.T) {
	buf := []byte{0, 0, 0, 1}
	b58 := CheckEncode(buf)
	dec, err := CheckDecode(b58)
	require.NoError(t, err)
	require.Equal(t, buf, dec)
}
```

File: pkg/encoding/address/address.go [hunk 2]
```diff
@@ -1,21 +1,20 @@
...
 import (
 	"github.com/CityOfZion/neo-go/pkg/encoding/base58"
 	"github.com/CityOfZion/neo-go/pkg/util"
 )
 
-// AddressFromUint160 returns the "NEO address" from the given
-// Uint160.
-func AddressFromUint160(u util.Uint160) string {
+// EncodeUint160 returns the "NEO address" from the given Uint160.
+func EncodeUint160(u util.Uint160) string {
```
Comment (fyrchik) [inline pkg/encoding/address/address.go]: May be call it `ToString` or `FromUint160` ?
Looking at previous files it was not completely obvious what `EncodeUint160` it does.

File: pkg/encoding/address/address.go [hunk 3]
```diff
@@ -1,21 +1,20 @@
...
 	// Dont forget to prepend the Address version 0x17 (23) A
 	b := append([]byte{0x17}, u.BytesBE()...)
 	return base58.CheckEncode(b)
 }
 
-// Uint160DecodeAddress attempts to decode the given NEO address string
+// DecodeUint160 attempts to decode the given NEO address string
 // into an Uint160.
-func Uint160DecodeAddress(s string) (u util.Uint160, err error) {
+func DecodeUint160(s string) (u util.Uint160, err error) {
```
Comment (fyrchik) [inline pkg/encoding/address/address.go]: Same here, I'd expect `DecodeUint160` to accept `Uint160` as _an argument_ and return some form of `address` (package name).
Example from stdlib: `hex.DecodeString(string) ([]byte error)`

File: pkg/encoding/address/address.go (hunk 2 above)
Comment (roman-khimov) [reply pkg/encoding/address/address.go]: `Uint160ToString`?

File: pkg/encoding/address/address.go (hunk 3 above)
Comment (roman-khimov) [reply pkg/encoding/address/address.go]: `StringToUint160`? The problem here is that `address` doesn't define any type of its own, so `ToString` and `FromString` might seem a little strange.

File: pkg/encoding/base58/base58.go (hunk 1 above)
Comment (roman-khimov) [reply pkg/encoding/base58/base58.go]: Thanks for explanation, this must be a remnant of internal base58 implementation then. Added a test.

**Pull request merged**